### DISCUSSION TITLE
Configuring Next.js for Proxy Interception Between Server and External API with Fiddler/Charles

### DISCUSSION THREAD

**Discussion by robertrutkowski** [vercel/next.js#62417] [Help]
### Summary

I'm developing a Next.js application and am in need of configuring my development setup to intercept and analyze the HTTP/HTTPS traffic that flows between the Next.js server-side logic (specifically route handlers) and an external API. My goal is to use interception tools like Fiddler or Charles Proxy to closely inspect the requests, including detailed headers and body content, that are exchanged between **my Next.js application's server-side** and **the external services** it communicates with.

**My questions are focused on the server-to-external API communication:**

1. How can I effectively configure Next.js so that it routes its server-side API requests through a local proxy server set up by either Fiddler or Charles?
2. What's the best approach to configuring such a proxy within Next.js, enabling a seamless transition between development and production environments without necessitating code changes?
3. Are there specific best practices or security/performance considerations to keep in mind when implementing this proxy setup for server-to-API communications within a Next.js context?

Despite attempting various methods, including environment variable configurations and direct proxy settings in my HTTP/HTTPS requests, I'm still facing challenges in capturing certain requests made from the server-side route handlers to external APIs.

Any guidance, insights, or sample code snippets that could assist me in properly establishing this interception setup would be immensely appreciated.

### Additional information

I tried setup custom server like in example below. But it doesn't work. I can't see in fiddler requests from server to my API (requests from browser to next server are visible of course).

### Example

```
const { createServer } = require("http");
const { parse } = require("url");
const next = require("next");
const { bootstrap } = require("global-agent");

const dev = process.env.NODE_ENV !== "production";
const hostname = "localhost";
const port = 3000;
const app = next({ dev, hostname, port });
const handle = app.getRequestHandler();

app.prepare().then(() => {
  bootstrap();

  createServer(async (req, res) => {
    try {
      const parsedUrl = parse(req.url, true);

      await handle(req, res, parsedUrl);
    } catch (err) {
      console.error("Error occurred handling", req.url, err);
      res.statusCode = 500;
      res.end("internal server error");
    }
  })
    .once("error", (err) => {
      console.error(err);
      process.exit(1);
    })
    .listen(port, () => {
      console.log(`> Ready on http://${hostname}:${port}`);
    });
});
```

**Comment by markdemich-fl** [comment]
@robertrutkowski , did you ever solve this?  I'm having a similar issue.  I posted a similar question here https://github.com/vercel/next.js/discussions/75964, but yours is much better.  From what I found out so far, global-agent has no affect on the native fetch in nodejs and that's what next.js wants us to use.  I tried a hacky solution to install undici and set the globalDispatcher, but I still could not get that to work.  I'm going to attempt again to see if I was making a silly mistake.  I really want to use fiddler.
